File: app/src/main/java/com/robifr/ledger/ui/product/ProductCardWideComponent.java
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.product;

import android.content.Context;
import android.view.View;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatDelegate;
import com.google.android.material.shape.ShapeAppearanceModel;
import com.robifr.ledger.R;
import com.robifr.ledger.data.model.ProductModel;
import com.robifr.ledger.databinding.ProductCardWideBinding;
import com.robifr.ledger.util.CurrencyFormat;
import java.math.BigDecimal;
import java.util.Objects;

public class ProductCardWideComponent {
  @NonNull private final Context _context;
  @NonNull private final ProductCardWideBinding _binding;

  public ProductCardWideComponent(
      @NonNull Context context, @NonNull ProductCardWideBinding binding) {
    this._context = Objects.requireNonNull(context);
    this._binding = Objects.requireNonNull(binding);

    final ShapeAppearanceModel imageShape =
        ShapeAppearanceModel.builder(
                this._context,
                com.google.android.material.R.style.Widget_MaterialComponents_ShapeableImageView,
                R.style.Shape_Card)
            .build();
    this._binding.normalCard.image.shapeableImage.setShapeAppearanceModel(imageShape);
    this._binding.expandedCard.image.shapeableImage.setShapeAppearanceModel(imageShape);
  }

  public void setNormalCardProduct(@NonNull ProductModel product) {
    Objects.requireNonNull(product);

    this._setId(product.id(), true);
    this._setName(product.name(), true);
    this._setPrice(product.price(), true);
  }

  public void setExpandedCardProduct(@NonNull ProductModel product) {
    Objects.requireNonNull(product);

    this._setId(product.id(), false);
    this._setName(product.name(), false);
    this._setPrice(product.price(), false);
  }

  public void setCardExpanded(boolean isExpanded) {
    final int normalCardVisibility = isExpanded ? View.GONE : View.VISIBLE;
    final int expandedCardVisibility = isExpanded ? View.VISIBLE : View.GONE;

    this._binding.normalCard.getRoot().setVisibility(normalCardVisibility);
    this._binding.expandedCard.getRoot().setVisibility(expandedCardVisibility);
  }

  public void setCardChecked(boolean isChecked) {
    final int textVisibility = isChecked ? View.GONE : View.VISIBLE;
    final int iconVisibility = isChecked ? View.VISIBLE : View.GONE;

    this._binding.cardView.setChecked(isChecked);
    this._binding.normalCard.image.text.setVisibility(textVisibility);
    this._binding.normalCard.image.icon.setVisibility(iconVisibility);
    this._binding.expandedCard.image.text.setVisibility(textVisibility);
    this._binding.expandedCard.image.icon.setVisibility(iconVisibility);
  }

  public void reset() {
    this._binding.normalCard.uniqueId.setText(null);
    this._binding.normalCard.uniqueId.setEnabled(false);
    this._binding.expandedCard.uniqueId.setText(null);
    this._binding.expandedCard.uniqueId.setEnabled(false);

    this._binding.cardView.setChecked(false);
    this._binding.normalCard.name.setText(null);
    this._binding.normalCard.image.text.setText(null);
    this._binding.normalCard.image.icon.setVisibility(View.GONE);
    this._binding.expandedCard.name.setText(null);
    this._binding.expandedCard.image.text.setText(null);
    this._binding.expandedCard.image.icon.setVisibility(View.GONE);

    this._binding.normalCard.price.setText(null);
    this._binding.expandedCard.price.setText(null);
  }

  private void _setId(@Nullable Long id, boolean isNormalCard) {
    final boolean isIdExists = id != null;
    final String productId =
        isIdExists ? id.toString() : this._context.getString(R.string.symbol_notavailable);

    if (isNormalCard) {
      this._binding.normalCard.uniqueId.setText(productId);
      this._binding.normalCard.uniqueId.setEnabled(isIdExists);
    } else {
      this._binding.expandedCard.uniqueId.setText(productId);
      this._binding.expandedCard.uniqueId.setEnabled(isIdExists);
    }
  }

  private void _setName(@NonNull String name, boolean isNormalCard) {
    Objects.requireNonNull(name);

    final String initialLetterName = name.trim().substring(0, Math.min(1, name.trim().length()));

    if (isNormalCard) {
      this._binding.normalCard.name.setText(name);
      this._binding.normalCard.image.text.setText(initialLetterName);
    } else {
      this._binding.expandedCard.name.setText(name);
      this._binding.expandedCard.image.text.setText(initialLetterName);
    }
  }

  private void _setPrice(long price, boolean isNormalCard) {
    final String formattedPrice =
        CurrencyFormat.format(
            BigDecimal.valueOf(price), AppCompatDelegate.getApplicationLocales().toLanguageTags());

    if (isNormalCard) this._binding.normalCard.price.setText(formattedPrice);
    else this._binding.expandedCard.price.setText(formattedPrice);
  }
}

File: app/src/main/java/io/github/robifr/ledger/components/ProductCardWideComponent.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.robifr.ledger.components

import android.content.Context
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.view.isGone
import androidx.core.view.isVisible
import com.google.android.material.shape.ShapeAppearanceModel
import io.github.robifr.ledger.R
import io.github.robifr.ledger.data.model.ProductModel
import io.github.robifr.ledger.databinding.ProductCardWideBinding
import io.github.robifr.ledger.util.CurrencyFormat

class ProductCardWideComponent(
    private val _context: Context,
    private val _binding: ProductCardWideBinding
) {
  init {
    val imageShape: ShapeAppearanceModel =
        ShapeAppearanceModel.builder()
            .setAllCornerSizes(_context.resources.getDimension(R.dimen.corner_small))
            .build()
    _binding.normalCard.image.shapeableImage.shapeAppearanceModel = imageShape
    _binding.expandedCard.image.shapeableImage.shapeAppearanceModel = imageShape
  }

  fun setNormalCardProduct(product: ProductModel) {
    _setId(product.id, true)
    _setName(product.name, true)
    _setPrice(product.price, true)
  }

  fun setExpandedCardProduct(product: ProductModel) {
    _setId(product.id, false)
    _setName(product.name, false)
    _setPrice(product.price, false)
  }

  fun setCardExpanded(isExpanded: Boolean) {
    _binding.normalCard.root.isVisible = !isExpanded
    _binding.expandedCard.root.isVisible = isExpanded
  }

  fun setCardChecked(isChecked: Boolean) {
    _binding.cardView.isChecked = isChecked
    _binding.normalCard.image.text.isVisible = !isChecked
    _binding.normalCard.image.icon.isVisible = isChecked
    _binding.expandedCard.image.text.isVisible = !isChecked
    _binding.expandedCard.image.icon.isVisible = isChecked
  }

  fun reset() {
    _binding.normalCard.uniqueId.text = null
    _binding.normalCard.uniqueId.isEnabled = false
    _binding.expandedCard.uniqueId.text = null
    _binding.expandedCard.uniqueId.isEnabled = false

    _binding.cardView.isChecked = false
    _binding.normalCard.name.text = null
    _binding.normalCard.image.text.text = null
    _binding.normalCard.image.icon.isGone = true
    _binding.expandedCard.name.text = null
    _binding.expandedCard.image.text.text = null
    _binding.expandedCard.image.icon.isGone = true

    _binding.normalCard.price.text = null
    _binding.expandedCard.price.text = null
  }

  private fun _setId(id: Long?, isNormalCard: Boolean) {
    val customerId: String = id?.toString() ?: _context.getString(R.string.symbol_notAvailable)
    if (isNormalCard) {
      _binding.normalCard.uniqueId.text = customerId
      _binding.normalCard.uniqueId.isEnabled = id != null
    } else {
      _binding.expandedCard.uniqueId.text = customerId
      _binding.expandedCard.uniqueId.isEnabled = id != null
    }
  }

  private fun _setName(name: String, isNormalCard: Boolean) {
    if (isNormalCard) {
      _binding.normalCard.name.text = name
      _binding.normalCard.image.text.text = name.take(1)
    } else {
      _binding.expandedCard.name.text = name
      _binding.expandedCard.image.text.text = name.take(1)
    }
  }

  private fun _setPrice(price: Long, isNormalCard: Boolean) {
    val formattedPrice: String =
        CurrencyFormat.formatCents(
            price.toBigDecimal(), AppCompatDelegate.getApplicationLocales().toLanguageTags())
    if (isNormalCard) _binding.normalCard.price.text = formattedPrice
    else _binding.expandedCard.price.text = formattedPrice
  }
}
